### DISCUSSION TITLE
[v2] More specific __eq__ and __hash__ for BaseModel

### DISCUSSION THREAD

**Discussion by AlexanderPodorov** [pydantic/pydantic#6717] [Pydantic V2]
I've found myself writing something like this:

```python
from pydantic import BaseModel
from typing import Self

class Test(BaseModel):
    id: int
    payload: str

    def __eq__(self, other: Self) -> bool:
        return self.id == other.id

    def __hash__(self) -> int:
        return hash(self.id)
```

In my use case if two objects have the same `id` they are considered equal, no matter which `payload` value is. This is particularly useful having those objects in a container and be able to throw validation error if a container contains duplicate items.
This approach generally works, however I'm wondering if there's a possibility to reduce the boilerplate, e.g. to automatically generate `__eq__` and `__hash__` just on certain fields?

E.g. it is possible for stdlib `dataclasses`:
```python
from dataclasses import dataclass, field

@dataclass(frozen=True)
class Test:
    id: int = field(compare=True)
    payload: str = field(compare=False)
```
In this case the dataclass will generate `__eq__` and `__hash__` properly.
It would be great to have something like that in `pydantic v2`. E.g. we could have the `eq` parameter of `Field` to achieve the same behavior.

Suggested code:
```python
from pydantic import BaseModel, Field
from typing import Annotated

class Test(BaseModel, frozen=True):
    id: Annotated[int, Field(compare=True)]
    payload: Annotated[str, Field(compare=False)]
```
If `compare=True` by default in the `Field` this change would be backward compatible. Thanks!

**Comment by miikka** [comment]
We're facing a problem where this would help. We have got some hand-written memoization code that we're upgrading from Pydantic V1 to V2. Here's a simplified example:

```python
from typing import Optional

from pydantic import BaseModel, ConfigDict, PrivateAttr


class MyModel(BaseModel):
    model_config = ConfigDict(frozen=True)

    _cached_value: Optional[str] = PrivateAttr(None)

    def _calculate_value(self):
        print("Calculating value...")
        return "hello"

    def get_value(self):
        if self._cached_value is None:
            self._cached_value = self._calculate_value()
        return self._cached_value

instance1 = MyModel()
instance2 = MyModel()
assert instance1 == instance2, "instances should be equal before getting value"

print(f"{instance1.get_value()=}")
assert instance1 == instance2, "instances should be equal after getting value"
```

In Pydantic 1.10.2, this works just fine. In Pydantic 2.1.1, the latter assertion fails because the private attribute `_cached_value` is taken into consideration when comparing the model instances. If we could set `compare=False` for `_cached_value`, it would work the way we intended.

**Comment by sambrilleman** [comment]
Is there any update on this? Is something similar to `field(compare = False)` likely to be supported? As far as I can see from the docs, there isn't any mention of functionality like this for `BaseModel` only `dataclass` ?

**Comment by bitclick** [comment]
I am new to pydantic, coming from `dataclasses` and i wonder "what am i missing?".. Where is the compare-functionality?   (also/offtopic: why distinguish between `dataclasses` and `basemodel` at all in pydantic?)

as it is, `BaseModel` seems like a mixed bag instead of an upgrade from `dataclasses`

**Comment by apiwat-chantawibul** [comment]
Here is my hack --- wrap private field by an `AlwaysEqual` class that equates all instances of the same type.

Only works for private attribute because I have not implemented `__get_pydantic_core_schema__` for the class. But this is enough for my use case.

```python
import pytest
from pydantic import (
    BaseModel,
    ConfigDict,
)


class AlwaysEqual[T]:
    """Used for excluding pydantic private attribute from being used in equality checks"""
    def __init__(self, value: T):
        self.value = value

    def __eq__(self, other):
        if not isinstance(other, AlwaysEqual):
            return NotImplemented
        return True

    def __hash__(self):
        return hash(self.__class__)

    def __repr__(self):
        return f'{self.__class__.__name__}({self.value!r})'


def test_eq():
    one = AlwaysEqual(1)
    two = AlwaysEqual(2)
    assert one == two
    assert one.value != two.value


def test_hash():
    one = AlwaysEqual(1)
    two = AlwaysEqual(2)
    d = { one, two }
    assert len(d) == 1


def test_pydantic_model_equality_ignore_always_equal_private_field():
    class Item(BaseModel):
        name: str
        _created_by: AlwaysEqual[str]

        @classmethod
        def create(cls, by: str, /, **kwargs):
            item = cls(**kwargs)
            item._created_by = AlwaysEqual(by)
            return item

    one = Item.create('one', name = 'Martha')
    two = Item.create('two', name = 'Martha')
    assert one == two
    assert one._created_by == two._created_by
    assert one._created_by.value != two._created_by.value

    different = Item.create('one', name = 'Matilda')
    assert one != different


def test_pydantic_model_hash_ignore_always_equal_private_field():
    class Item(BaseModel):
        model_config = ConfigDict(frozen = True)
        name: str
        _created_by: AlwaysEqual[str]

        @classmethod
        def create(cls, by: str, /, **kwargs):
            item = cls(**kwargs)
            item._created_by = AlwaysEqual(by)
            return item

    one = Item.create('one', name = 'Martha')
    two = Item.create('two', name = 'Martha')
    different = Item.create('one', name = 'Matilda')
    assert hash(one) == hash(two)
    assert hash(one) != hash(different)
```
